File: Q001-050/Q050.R
# encoding=utf=8
# Euler 50: Consecutive prime sum

source('prime.R')

binarySearch <- function(vec, value) {
  begin <- 1L
  end <- length(vec)
  repeat {
    center <- as.integer((begin+end)/2)
    if(center==begin) {
      if(vec[begin]==value)
        return(begin)
      else if(vec[end]==value)
        return(end)
      else
        return(0)
    } else if(value>vec[center]) {
      begin <- center
    } else {
      end <- center
    }
  }
}

### Main:
library(plyr)
limit <- 1000000
result <- 0
numberOfPrimes <- 0L
primes <- get.prime(1, limit)
primeSum <- c(0, cumsum(primes))

progress <- create_progress_bar('text')
print(system.time({
  progress$init(length(primeSum))
  for(i in (numberOfPrimes+1):length(primeSum)) {
    for(j in (i-numberOfPrimes):1) {
      d <- primeSum[i] - primeSum[j]
      if(d>limit) break
      if(binarySearch(primes,d)>0) {
        numberOfPrimes <- i - j
        result <- d
        ri <- i
        rj <- j
      }
    }
    progress$step()
  }
  progress$term()
}))
cat(sprintf('%d has %d primes.\n', result, numberOfPrimes))

File: Q001-050/Q050.py
#!/usr/bin/env python
# encoding=utf-8

import sys
import time
sys.path.append('..')
from utils.Prime import isPrime, getPrime
from utils.search import binarySearch

limit = 1000000L
result = 0
numberOfPrimes = 0L

primes = getPrime(1L,limit)

primeSum = [0]
for pm in primes:
  primeSum += [primeSum[-1]+pm]

start_time = time.time()
for i in xrange(numberOfPrimes,len(primeSum)):
  for j in xrange(i-(numberOfPrimes+1),-1,-1):
    d = primeSum[i] - primeSum[j]
    if d > limit: break
    if binarySearch(primes,d)>=0:
      numberOfPrimes = i - j
      result = d
      ri = i
      rj = j
end_time = time.time()
print 'Time elapsed:', end_time-start_time
print result, 'has', numberOfPrimes, 'primes.'

